File: samples_reference_c++/ParametrizeCamera_Configurations/ParametrizeCamera_Configurations.cpp
// ParametrizeCamera_Configurations.cpp
/*
    Note: Before getting started, Basler recommends reading the "Programmer's Guide" topic
    in the pylon C++ API documentation delivered with pylon.
    If you are upgrading to a higher major version of pylon, Basler also
    strongly recommends reading the "Migrating from Previous Versions" topic in the pylon C++ API documentation.

    The instant camera allows to install event handlers for configuration purposes
    and for handling the grab results. This is very useful for handling standard
    camera setups and image processing tasks.

    This sample shows how to use configuration event handlers by applying the standard
    configurations and registering sample configuration event handlers.

    Configuration event handlers are derived from the CConfigurationEventHandler base class.
    CConfigurationEventHandler provides virtual methods that can be overridden. If the
    configuration event handler is registered these methods are called when the state of the
    instant camera objects changes, e.g. when the camera object is opened or closed.

    The standard configuration event handlers override the OnOpened method. The overridden method
    parametrizes the camera.

    Device specific camera classes, e.g. for GigE cameras, provide specialized
    event handler base classes, e.g. CBaslerGigEConfigurationEventHandler.
*/

// Include files to use the pylon API.
#include <pylon/PylonIncludes.h>

// Include files used by samples.
#include "../include/ImageEventPrinter.h"
#include "../include/ConfigurationEventPrinter.h"
#include "../include/PixelFormatAndAoiConfiguration.h"

// Namespace for using pylon objects.
using namespace Pylon;

// Namespace for using cout.
using namespace std;

// Number of images to be grabbed.
static const uint32_t c_countOfImagesToGrab = 3;


int main( int /*argc*/, char* /*argv*/[] )
{
    // The exit code of the sample application.
    int exitCode = 0;

    // Before using any pylon methods, the pylon runtime must be initialized.
    PylonInitialize();

    try
    {
        // Create an instant camera object with the first camera device found.
        CInstantCamera camera( CTlFactory::GetInstance().CreateFirstDevice() );

        // Print the model name of the camera.
        cout << "Using device: " << camera.GetDeviceInfo().GetModelName() << endl << endl;

        // For demonstration purposes only, register an image event handler.
        // printing out information about the grabbed images.
        camera.RegisterImageEventHandler( new CImageEventPrinter, RegistrationMode_Append, Cleanup_Delete );

        // This smart pointer will receive the grab result data.
        CGrabResultPtr ptrGrabResult;



        cout << "Grab using continuous acquisition:" << endl << endl;

        // Register the standard configuration event handler for setting up the camera for continuous acquisition.
        // By setting the registration mode to RegistrationMode_ReplaceAll, the new configuration handler replaces the
        // default configuration handler that has been automatically registered when creating the
        // instant camera object.
        // The handler is automatically deleted when deregistered or when the registry is cleared if Cleanup_Delete is specified.
        camera.RegisterConfiguration( new CAcquireContinuousConfiguration, RegistrationMode_ReplaceAll, Cleanup_Delete );

        // The camera's Open() method calls the configuration handler's OnOpened() method that
        // applies the required parameter modifications.
        camera.Open();

        // The registered configuration event handler has done its parametrization now.
        // Additional parameters could be set here.

        // Grab some images for demonstration.
        camera.StartGrabbing( c_countOfImagesToGrab );
        while (camera.IsGrabbing())
        {
            camera.RetrieveResult( 5000, ptrGrabResult, TimeoutHandling_ThrowException );
        }

        // Close the camera.
        camera.Close();



        cout << "Grab using software trigger mode:" << endl << endl;

        // Register the standard configuration event handler for setting up the camera for software
        // triggering.
        // The current configuration is replaced by the software trigger configuration by setting the
        // registration mode to RegistrationMode_ReplaceAll.
        camera.RegisterConfiguration( new CSoftwareTriggerConfiguration, RegistrationMode_ReplaceAll, Cleanup_Delete );

        // StartGrabbing() calls the camera's Open() automatically if the camera is not open yet.
        // The Open method calls the configuration handler's OnOpened() method that
        // sets the required parameters for enabling software triggering.

        // Grab some images for demonstration.
        camera.StartGrabbing( c_countOfImagesToGrab );
        while (camera.IsGrabbing())
        {
            // Execute the software trigger. The call waits up to 1000 ms for the camera
            // to be ready to be triggered.
            camera.WaitForFrameTriggerReady( 1000, TimeoutHandling_ThrowException );
            camera.ExecuteSoftwareTrigger();
            camera.RetrieveResult( 5000, ptrGrabResult, TimeoutHandling_ThrowException );
        }
        // StopGrabbing() is called from RetrieveResult if the number of images
        // to grab has been reached. Since the camera was opened by StartGrabbing()
        // it is closed by StopGrabbing().

        // The CSoftwareTriggerConfiguration, like all standard configurations, is provided as a header file.
        // The source code can be copied and modified to meet application specific needs, e.g.
        // the CSoftwareTriggerConfiguration class could easily be changed into a hardware trigger configuration.



        cout << "Grab using single frame acquisition:" << endl << endl;

        // Register the standard configuration event handler for configuring single frame acquisition.
        // The previous configuration is removed by setting the registration mode to RegistrationMode_ReplaceAll.
        camera.RegisterConfiguration( new CAcquireSingleFrameConfiguration, RegistrationMode_ReplaceAll, Cleanup_Delete );

        // GrabOne calls StartGrabbing and StopGrabbing internally.
        // As seen above Open() is called by StartGrabbing and
        // the OnOpened() method of the CAcquireSingleFrameConfiguration handler is called.
        camera.GrabOne( 5000, ptrGrabResult );

        // To continuously grab single images it is much more efficient to open the camera before grabbing.
        // Note: The software trigger mode (see above) should be used for grabbing single images if you want to maximize frame rate.

        // Now, the camera parameters are applied in the OnOpened method of the configuration object.
        camera.Open();

        // Additional parameters could be set here.

        // Grab some images for demonstration.
        camera.GrabOne( 5000, ptrGrabResult );
        camera.GrabOne( 5000, ptrGrabResult );
        camera.GrabOne( 5000, ptrGrabResult );

        // Close the camera.
        camera.Close();



        cout << "Grab using multiple configuration objects:" << endl << endl;

        // Register the standard event handler for configuring single frame acquisition.
        // The current configuration is replaced by setting the registration mode to RegistrationMode_ReplaceAll.
        camera.RegisterConfiguration( new CAcquireSingleFrameConfiguration, RegistrationMode_ReplaceAll, Cleanup_Delete );

        // Register an additional configuration handler to set the image format and adjust the AOI.
        // By setting the registration mode to RegistrationMode_Append, the configuration handler is added instead of replacing
        // the already registered configuration handler.
        camera.RegisterConfiguration( new CPixelFormatAndAoiConfiguration, RegistrationMode_Append, Cleanup_Delete );

        // Create an event printer on the heap.
        CConfigurationEventPrinter* pEventPrinterObject = new CConfigurationEventPrinter;
        // Register the handler object and define Cleanup_None so that it is not deleted by the camera object.
        // It must be ensured, that the configuration handler "lives" at least until the handler is deregistered!
        camera.RegisterConfiguration( pEventPrinterObject, RegistrationMode_Append, Cleanup_None );

        // Grab an image for demonstration. Configuration events are printed.
        cout << endl << "Grab, configuration events are printed:" << endl << endl;
        camera.GrabOne( 5000, ptrGrabResult );

        // Deregister the event handler.
        camera.DeregisterConfiguration( pEventPrinterObject );
        // The event handler can now be deleted.
        delete pEventPrinterObject;
        pEventPrinterObject = NULL;

        // Grab an image for demonstration. Configuration events are not printed.
        cout << endl << "Grab, configuration events are not printed:" << endl << endl;
        camera.GrabOne( 5000, ptrGrabResult );
    }
    catch (const GenericException& e)
    {
        // Error handling.
        cerr << "An exception occurred." << endl
            << e.GetDescription() << endl;
        exitCode = 1;
    }

    // Comment the following two lines to disable waiting on exit
    cerr << endl << "Press enter to exit." << endl;
    while (cin.get() != '\n');

    // Releases all pylon resources.
    PylonTerminate();

    return exitCode;
}

File: samples_reference_pypylon/parametrize_camera_configurations/parametrize_camera_configurations.py
#!/usr/bin/env python3
"""
parametrize_camera_configurations.py

This sample demonstrates how to use configuration event handlers by applying standard
configurations and registering custom configuration event handlers.

Configuration event handlers are derived from the ConfigurationEventHandler base class
and provide methods that are called when the state of the instant camera object changes,
e.g. when the camera object is opened or closed.

This is equivalent to samples_reference_c++/ParametrizeCamera_Configurations/
"""

from typing import Optional
import pypylon.pylon as pylon
from pypylon import genicam


class CustomConfigurationEventHandler(pylon.ConfigurationEventHandler):
    """Custom configuration event handler that prints information about camera state changes."""
    
    def __init__(self, name: str):
        super().__init__()
        self.name = name
    
    def OnOpened(self, camera: pylon.InstantCamera):
        """Called when the camera is opened."""
        print(f"[{self.name}] OnOpened: Camera opened")
    
    def OnClosed(self, camera: pylon.InstantCamera):
        """Called when the camera is closed."""
        print(f"[{self.name}] OnClosed: Camera closed")
    
    def OnGrabStart(self, camera: pylon.InstantCamera):
        """Called when grabbing starts."""
        print(f"[{self.name}] OnGrabStart: Grabbing started")
    
    def OnGrabStop(self, camera: pylon.InstantCamera):
        """Called when grabbing stops."""
        print(f"[{self.name}] OnGrabStop: Grabbing stopped")


class PixelFormatAndAoiConfiguration(pylon.ConfigurationEventHandler):
    """Custom configuration handler that sets pixel format and AOI."""
    
    def OnOpened(self, camera: pylon.InstantCamera):
        """Configure pixel format and AOI when camera is opened."""
        try:
            print("[PixelFormatAndAoiConfiguration] Configuring pixel format and AOI...")
            
            # Try to set offsets to minimum
            try:
                if hasattr(camera, 'OffsetX') and camera.OffsetX.IsWritable():
                    camera.OffsetX.Value = camera.OffsetX.Min
                if hasattr(camera, 'OffsetY') and camera.OffsetY.IsWritable():
                    camera.OffsetY.Value = camera.OffsetY.Min
            except (genicam.GenericException, AttributeError):
                pass
            
            # Try to maximize width and height
            try:
                if hasattr(camera, 'Width') and camera.Width.IsWritable():
                    camera.Width.Value = camera.Width.Max
                if hasattr(camera, 'Height') and camera.Height.IsWritable():
                    camera.Height.Value = camera.Height.Max
            except (genicam.GenericException, AttributeError):
                pass
            
            # Try to set pixel format to Mono8
            try:
                if hasattr(camera, 'PixelFormat') and camera.PixelFormat.IsWritable():
                    camera.PixelFormat.Value = "Mono8"
            except (genicam.GenericException, AttributeError):
                pass
            
            print("[PixelFormatAndAoiConfiguration] Configuration applied successfully")
            
        except Exception as e:
            print(f"[PixelFormatAndAoiConfiguration] Error during configuration: {e}")


class SimpleImageEventHandler(pylon.ImageEventHandler):
    """Simple image event handler that prints image information."""
    
    def OnImageGrabbed(self, camera: pylon.InstantCamera, grab_result: pylon.GrabResult):
        """Called when an image is grabbed."""
        if grab_result.GrabSucceeded():
            print(f"[ImageEventHandler] Image grabbed successfully: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
        else:
            print(f"[ImageEventHandler] Image grab failed: {grab_result.GetErrorDescription()}")


def demonstrate_continuous_acquisition(camera: pylon.InstantCamera) -> None:
    """Demonstrate continuous acquisition with default configuration."""
    print("=" * 60)
    print("Grab using continuous acquisition:")
    print("=" * 60)
    
    # For continuous acquisition, we can use the default configuration
    # or register a custom configuration handler
    
    # Register a custom configuration handler for demonstration
    config_handler = CustomConfigurationEventHandler("ContinuousAcquisition")
    camera.RegisterConfiguration(config_handler, pylon.RegistrationMode_ReplaceAll, pylon.Cleanup_Delete)
    
    # Open the camera (this will trigger OnOpened)
    camera.Open()
    
    # Grab some images
    images_to_grab = 3
    camera.StartGrabbingMax(images_to_grab)
    
    while camera.IsGrabbing():
        grab_result = camera.RetrieveResult(5000, pylon.TimeoutHandling_ThrowException)
        if grab_result.GrabSucceeded():
            print(f"  Grabbed image: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
        grab_result.Release()
    
    # Close the camera (this will trigger OnClosed)
    camera.Close()
    print()


def demonstrate_software_trigger(camera: pylon.InstantCamera) -> None:
    """Demonstrate software trigger configuration."""
    print("=" * 60)
    print("Grab using software trigger mode:")
    print("=" * 60)
    
    # Register the standard configuration event handler for software triggering
    camera.RegisterConfiguration(pylon.SoftwareTriggerConfiguration(), 
                                pylon.RegistrationMode_ReplaceAll, 
                                pylon.Cleanup_Delete)
    
    # Add a custom configuration handler for demonstration
    config_handler = CustomConfigurationEventHandler("SoftwareTrigger")
    camera.RegisterConfiguration(config_handler, pylon.RegistrationMode_Append, pylon.Cleanup_Delete)
    
    # Open the camera (configurations will be applied)
    camera.Open()
    
    # Check if camera can be triggered
    if camera.CanWaitForFrameTriggerReady():
        # Grab some images using software trigger
        images_to_grab = 3
        camera.StartGrabbingMax(images_to_grab)
        
        while camera.IsGrabbing():
            # Wait for camera to be ready for next trigger
            camera.WaitForFrameTriggerReady(1000, pylon.TimeoutHandling_ThrowException)
            
            # Execute software trigger
            camera.ExecuteSoftwareTrigger()
            
            # Retrieve the result
            grab_result = camera.RetrieveResult(5000, pylon.TimeoutHandling_ThrowException)
            if grab_result.GrabSucceeded():
                print(f"  Triggered image: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
            grab_result.Release()
    else:
        print("  Camera does not support software triggering")
    
    camera.Close()
    print()


def demonstrate_single_frame_acquisition(camera: pylon.InstantCamera) -> None:
    """Demonstrate single frame acquisition configuration."""
    print("=" * 60)
    print("Grab using single frame acquisition:")
    print("=" * 60)
    
    # Register the standard configuration handler for single frame acquisition
    camera.RegisterConfiguration(pylon.AcquireSingleFrameConfiguration(), 
                                pylon.RegistrationMode_ReplaceAll, 
                                pylon.Cleanup_Delete)
    
    # Add a custom configuration handler for demonstration
    config_handler = CustomConfigurationEventHandler("SingleFrame")
    camera.RegisterConfiguration(config_handler, pylon.RegistrationMode_Append, pylon.Cleanup_Delete)
    
    # Use GrabOne for single frame acquisition
    grab_result = camera.GrabOne(5000)
    if grab_result.GrabSucceeded():
        print(f"  Single frame grabbed: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
    grab_result.Release()
    
    # For continuous single frame grabbing, it's more efficient to open the camera first
    camera.Open()
    
    # Grab multiple single frames
    for i in range(3):
        grab_result = camera.GrabOne(5000)
        if grab_result.GrabSucceeded():
            print(f"  Single frame {i+1}: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
        grab_result.Release()
    
    camera.Close()
    print()


def demonstrate_multiple_configurations(camera: pylon.InstantCamera) -> None:
    """Demonstrate using multiple configuration handlers."""
    print("=" * 60)
    print("Grab using multiple configuration objects:")
    print("=" * 60)
    
    # Register single frame acquisition configuration
    camera.RegisterConfiguration(pylon.AcquireSingleFrameConfiguration(), 
                                pylon.RegistrationMode_ReplaceAll, 
                                pylon.Cleanup_Delete)
    
    # Register additional configuration handler for pixel format and AOI
    camera.RegisterConfiguration(PixelFormatAndAoiConfiguration(), 
                                pylon.RegistrationMode_Append, 
                                pylon.Cleanup_Delete)
    
    # Register image event handler for demonstration
    camera.RegisterImageEventHandler(SimpleImageEventHandler(), 
                                   pylon.RegistrationMode_Append, 
                                   pylon.Cleanup_Delete)
    
    # Create a configuration event handler that we'll deregister later
    temp_config_handler = CustomConfigurationEventHandler("Temporary")
    camera.RegisterConfiguration(temp_config_handler, pylon.RegistrationMode_Append, pylon.Cleanup_None)
    
    print("\nGrab with all configuration handlers active:")
    grab_result = camera.GrabOne(5000)
    if grab_result.GrabSucceeded():
        print(f"  Image with all configurations: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
    grab_result.Release()
    
    # Deregister the temporary configuration handler
    camera.DeregisterConfiguration(temp_config_handler)
    del temp_config_handler
    
    print("\nGrab with temporary configuration handler removed:")
    grab_result = camera.GrabOne(5000)
    if grab_result.GrabSucceeded():
        print(f"  Image without temporary config: {grab_result.GetWidth()}x{grab_result.GetHeight()}")
    grab_result.Release()
    
    print()


def main() -> int:
    """
    Main function demonstrating configuration event handlers.
    
    Returns:
        int: Exit code (0 for success, 1 for error)
    """
    exit_code = 0
    
    try:
        # Initialize pylon runtime
        
        
        # Create an instant camera object with the first camera device found
        camera = pylon.InstantCamera(pylon.TlFactory.GetInstance().CreateFirstDevice())
        
        # Print the model name of the camera
        print(f"Using device: {camera.GetDeviceInfo().GetModelName()}\n")
        
        # Demonstrate different configuration approaches
        demonstrate_continuous_acquisition(camera)
        demonstrate_software_trigger(camera)
        demonstrate_single_frame_acquisition(camera)
        demonstrate_multiple_configurations(camera)
        
        print("Configuration event handlers demonstration completed successfully!")
        
    except genicam.GenericException as e:
        print(f"An exception occurred: {e}")
        exit_code = 1
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        exit_code = 1
    finally:
        # Ensure pylon resources are released
        
    
    return exit_code


if __name__ == "__main__":
    import sys
    
    print("=== Configuration Event Handlers Sample ===")
    print("This sample demonstrates various configuration event handlers and their usage.")
    print("Configuration handlers are called when camera state changes occur.\n")
    
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        print("\nInterrupted by user")
        
        sys.exit(1) 